src/plugins.ts

```diff
@@ -0,0 +1,95 @@
+import { SchemaBuilder, NexusAcceptedTypeDef } from "./builder";
+import { id } from "./utils";
+
+/**
+ * This is a read-only builder-like api exposed to plugins in the onInstall
+ * hook. It proxies the Nexus Builder which is a much bigger beast that we do
+ * not want to directly expose to plugins.
+ */
+type BuilderFacade = {
+  hasType: (typeName: string) => boolean;
+};
+
+/**
+ * This is the Neuxs Plugin interface that allows users to extend Nexus at
+ * particular extension points. Plugins are just functions that receive hooks
+ * which can then be registered upon with callbacks.
+ */
+export type Plugin = (hooks: Hooks) => void;
+
+type Hooks = {
+  onInstall: OnInstallHook;
+};
+
+type OnInstallHook = (onInstallHandler: OnInstallHandler) => void;
+
+type OnInstallHandler = (
+  builder: BuilderFacade
+) => {
+  types: NexusAcceptedTypeDef[];
+};
+
+/**
+ * This is the interface that Nexus uses to drive plugins meaning triggering the
+ * lifecycle events that plugins have registered for.
+ */
+type PluginController = {
+  triggerOnInstall: () => void;
+};
+
+/**
+ * This represents the internal data of a plugin controller.
+ */
+type PluginControllerState = {
+  onInstallHandler: OnInstallHandler;
+};
+
+/**
+ * Create new plugin controller state data with default values.
+ */
+const createControllerState = (): PluginControllerState => {
+  const defaultOnInstallHandler = () => ({
+    types: [],
+  });
+
+  return {
+    onInstallHandler: defaultOnInstallHandler,
+  };
+};
+
+/**
+ * Initialize a plugin. This will gather the hook handlers (aka. callbacks,
+ * event handlers) that the plugin has registered for. A controller is returned
+ * that permits Nexus to trigger the hooks so executing the hook handlers.
+ */
+export const initializePlugin = (
+  builder: SchemaBuilder,
+  plugin: Plugin
+): PluginController => {
+  const state = createControllerState();
+
+  plugin({
+    onInstall: (handler) => {
+      state.onInstallHandler = handler;
+    },
+  });
+
+  return {
+    triggerOnInstall: () => {
+      const { types } = state.onInstallHandler({
+        hasType: builder.hasType,
+      });
+      types.forEach(builder.addType);
```

Comment:
We should probably put some runtime checks here for better error handling. Especially helpful for JS.

Reply:
We might want to use runtime DSLs that permit deriving (inferring) TS static types:

* https://github.com/gcanti/io-ts
* https://github.com/jquense/yup
* https://github.com/pelotom/runtypes

Normally this approach would be used for the IO boundary of an app. But we can kind of think of JS Nexus users as being in this camp too, where all data they are passing into Nexus API requires runtime validation.

I would prefer if we could codegen runtime checks from our TS types but don't know of a good tool for that. Definitely seems the runtime-to-inference is the dominant approach. Fair since that approach requires zero special tooling.

Reply:
A goal I had with nexus was to have `graphql-js` as the only dependency of the lib  if possible (other than `tslib`, for obvious reasons). This is also the point of `addTypes`, the predicate checks are meant to guard here. 

This is also why I recommended having a container class for `Plugin`, following the patterns elsewhere so we can easily add a predicate guard.